include/pulsar/Client.h

```diff
@@ -66,6 +67,31 @@ class PULSAR_PUBLIC Client {
      */
     Client(const std::string& serviceUrl, const ClientConfiguration& clientConfiguration);
 
+    /**
+     * Create a Pulsar client object connecting to the specified cluster address and using the default
+     * configuration.
+     *
+     * <p>Instead of specifying a static service URL string (with {@link #serviceUrl(String)}), an application
+     * can pass a {@link ServiceUrlProvider} function that dynamically provide a service URL.
+     *
+     * @param serviceUrlProvider The serviceUrlProvider used to generate ServiceUrl.
+     * @throw std::invalid_argument if `serviceUrlProvider()` return a invalid url.
+     */
+    Client(ServiceUrlProvider serviceUrlProvider);
```

Comment:
Instead of adding new constructor overloads, we should add `setServiceUrlProvider` to `ClientConfiguration`.

Reply:
This request the user to choose between `Client(ServiceUrlProvider serviceUrlProvider);` and `Client(const std::string& serviceUrl);` constructors.

If add `setServiceUrlProvider` to `ClientConfiguration`. When the user sets it this way:

```c++
ClientConfiguration config;
config.setServiceUrlProvider(serviceUrlProvider);
Client client(serviceUrl);
```

This will be confuse.

Reply:
LGTM. Before this PR, `serviceUrl` is a required argument. But after this PR, only one of `serviceUrl` and `serviceUrlProvider` should be provided.